File: 621.任务调度器.cpp
/*
 * @lc app=leetcode.cn id=621 lang=cpp
 *
 * [621] 任务调度器
 */
#include <iostream>
#include <map>
#include <vector>
using namespace std;
// @lc code=start
class Solution {
private:
  inline char findMostChar(map<char, int> &hash) {
    auto mostChar = 'A';
    for (auto i = 'A'; i <= 'Z'; i++) {
      if (hash[i] > hash[mostChar]) {
        mostChar = i;
      }
    }
    return mostChar;
  }

public:
  /**
   * @brief
   * 71/71 cases passed (52 ms)
   * Your runtime beats 75.87 % of cpp submissions
   * Your memory usage beats 10.03 % of cpp submissions (33.7 MB)
   * @param tasks
   * @param n
   * @return int
   */
  int leastInterval(vector<char> &tasks, int n) {
    map<char, int> hash;
    for (auto task : tasks) {
      hash[task]++;
    }
    char mostChar;
    int length = 0, mostTasks = 0;
    for (; hash[mostChar = findMostChar(hash)];) {
      if (!mostTasks) {
        mostTasks = hash[mostChar];
        length = hash[mostChar] * (n + 1) - n;
      } else {
        if (hash[mostChar] == mostTasks) {
          length++;
        }
      }
      hash[mostChar] = 0;
    }
    return max(length, (int)tasks.size());
  }
};
// @lc code=end

File: 621.任务调度器.java
/*
 * @lc app=leetcode.cn id=621 lang=java
 *
 * [621] 任务调度器
 */
import java.util.HashMap;
import java.util.Map;

// @lc code=start
class Solution {

  private char findMostChar(Map<Character, Integer> hash) {
    var mostChar = 'A';
    for (var i = 'A'; i <= 'Z'; i++) {
      if (hash.getOrDefault(i, 0) > hash.getOrDefault(mostChar, 0)) {
        mostChar = i;
      }
    }
    return mostChar;
  }

  /**
   * 71/71 cases passed (14 ms)
   * Your runtime beats 38.26 % of java submissions
   * Your memory usage beats 91.65 % of java submissions (42 MB)
   * @param tasks
   * @param n
   * @return
   */
  public int leastInterval(char[] tasks, int n) {
    var hash = new HashMap<Character, Integer>();
    for (var task : tasks) {
      hash.put(task, hash.getOrDefault(task, 0) + 1);
    }
    char mostChar;
    int length = 0, mostTasks = 0;
    for (; hash.get(mostChar = findMostChar(hash)) != 0;) {
      if (mostTasks == 0) {
        mostTasks = hash.get(mostChar);
        length = hash.get(mostChar) * (n + 1) - n;
      } else {
        if (hash.get(mostChar) == mostTasks) {
          length++;
        }
      }
      hash.put(mostChar, 0);
    }
    return Math.max(length, tasks.length);
  }
}
// @lc code=end
